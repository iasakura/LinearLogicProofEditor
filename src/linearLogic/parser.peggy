{{
  const opToName = (op) => {
    if (op == '|') { return 'or'; }
    if (op == '@') { return 'par'; }
    if (op == '*') { return 'tensor'; }
    if (op == '&') { return 'and'; }
    if (op == '~') { return 'not'; }
    if (op == '!') { return 'ofCourse'; }
    if (op == '?') { return 'whiNot'; }
  }
}}

Formula
  = f1:Formula1 _ op:("|" / "@") _ f2:Formula {
      return { name: 'binary', op: opToName(op), children: [f1, f2] };
    } /
    f: Formula1 {
      return f;
    }

Formula1
  = f1:Formula2 _ op:("*" / "&") _ f2:Formula1 {
      return { name: 'binary', op: opToName(op), children: [f1, f2] };
    } /
    f: Formula2 {
      return f;
    }

Formula2
  = op:("~" / "!" / "?") _ f:Formula2 {
    return { name: 'unary', op: opToName(op), children: [f] };
  } /
  f:Formula3 {
    return f;
  }

Formula3
  = v:([a-zA-Z] [a-zA-Z0-9]*) {
    return {name: 'var', var: v[0] + v[1].join() };
  } / "(" _ f:Formula _ ")" {
    return f;
  }

_ "whitespace"
  = [ \t\n\r]*
