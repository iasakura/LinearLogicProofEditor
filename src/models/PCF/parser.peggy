{{
  import {Typ, Term} from '../models/PCF/types';
}}

Lam
  = "lam" _ v:Var _ ':' _ t:Type _ "." _ b:Lam {
      return { name: 'lambda', v: v, typ: t, body: b }
    } / c:Cond {
      return c;
    }

Cond
  = "if" _ c:App _ "then" _ th:App _ "else" _ el:App {
      return {name: 'app', terms: [{name: 'app', terms: [{name: 'app', terms: [{ name: "prim", v: 'cond' }, c]}, th]}, el]}
  } / a:App {
      return a;
  }

App
  = f:Prim args:(_ x:Prim)+ {
      return args.reduce((acc: Term, x: [string, Term]) => {
        return { name: 'app', terms: [acc, x[1]]}
      }, f);
  } / p:Prim {
      return p;
  }

Prim
   = v:Var {
       return {name: 'var', v: v }
   } / p:("succ" / "pred" / "iszero") {
       return {name: 'prim', v: p }
   } / n:Num {
       return {name: 'nat-const', val: n }
   } / b:Bool {
       return {name: 'bool-const', val: b }
   } / "(" _ e:Lam _ ")" {
       return e;
   }

Type
   = t1:InnerType ts:(_ "->" _ t2:InnerType)* {
       return [...ts.map((x: [string, string, string, Typ]) => x[3]).reverse(), t1].reduce((ty, t) => {
           return { name: 'arrow', types: [t, ty]}
       })
   } / InnerType

InnerType
    = 'nat' {
       return { name: 'nat' }
     } / 'bool' {
         return { name: 'bool' }
     } / "(" t:Type ")" {
         return t;
     }

Var
    = !Keywords v:Lit {
        return v;
    }

Lit
    = h:[a-zA-Z] t:[a-zA-Z0-9]* {
        return [h, ...t].join('');
}

Num
    = n:([0-9]+) {
        return parseInt(n, 10);
    }

Bool
    = b:("true" / "false") {
        return b === 'true';
    }

Keywords = "lam"/"if"/"then"/"else"/"iszero"/"succ"/"pred"

 _ "whitespace"
  = [ \t\n\r]*
