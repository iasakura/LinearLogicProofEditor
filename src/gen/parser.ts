
// Generated by peggy v. 1.2.0 (ts-pegjs plugin v. 1.2.1 )
//
// https://peggyjs.org/   https://github.com/metadevpro/ts-pegjs

"use strict";


  import { Formula } from '../models/linearLogic/Formula';

  const opToName = (op: string) => {
    if (op == '|') { return 'or'; }
    if (op == '@') { return 'par'; }
    if (op == '*') { return 'tensor'; }
    if (op == '&') { return 'and'; }
    if (op == '~') { return 'not'; }
    if (op == '!') { return 'ofCourse'; }
    if (op == '?') { return 'whyNot'; }
  }


export interface IFilePosition {
  offset: number;
  line: number;
  column: number;
}

export interface IFileRange {
  start: IFilePosition;
  end: IFilePosition;
  source: string;
}

export interface ILiteralExpectation {
  type: "literal";
  text: string;
  ignoreCase: boolean;
}

export interface IClassParts extends Array<string | IClassParts> {}

export interface IClassExpectation {
  type: "class";
  parts: IClassParts;
  inverted: boolean;
  ignoreCase: boolean;
}

export interface IAnyExpectation {
  type: "any";
}

export interface IEndExpectation {
  type: "end";
}

export interface IOtherExpectation {
  type: "other";
  description: string;
}

export type Expectation = ILiteralExpectation | IClassExpectation | IAnyExpectation | IEndExpectation | IOtherExpectation;

function peg$padEnd(str: string, targetLength: number, padString: string) {
  padString = padString || ' ';
  if (str.length > targetLength) {
    return str;
  }
  targetLength -= str.length;
  padString += padString.repeat(targetLength);
  return str + padString.slice(0, targetLength);
}

export class SyntaxError extends Error {
  public static buildMessage(expected: Expectation[], found: string | null) {
    function hex(ch: string): string {
      return ch.charCodeAt(0).toString(16).toUpperCase();
    }

    function literalEscape(s: string): string {
      return s
        .replace(/\\/g, "\\\\")
        .replace(/"/g,  "\\\"")
        .replace(/\0/g, "\\0")
        .replace(/\t/g, "\\t")
        .replace(/\n/g, "\\n")
        .replace(/\r/g, "\\r")
        .replace(/[\x00-\x0F]/g,            (ch) => "\\x0" + hex(ch) )
        .replace(/[\x10-\x1F\x7F-\x9F]/g, (ch) => "\\x"  + hex(ch) );
    }

    function classEscape(s: string): string {
      return s
        .replace(/\\/g, "\\\\")
        .replace(/\]/g, "\\]")
        .replace(/\^/g, "\\^")
        .replace(/-/g,  "\\-")
        .replace(/\0/g, "\\0")
        .replace(/\t/g, "\\t")
        .replace(/\n/g, "\\n")
        .replace(/\r/g, "\\r")
        .replace(/[\x00-\x0F]/g,            (ch) => "\\x0" + hex(ch) )
        .replace(/[\x10-\x1F\x7F-\x9F]/g, (ch) => "\\x"  + hex(ch) );
    }

    function describeExpectation(expectation: Expectation) {
      switch (expectation.type) {
        case "literal":
          return "\"" + literalEscape(expectation.text) + "\"";
        case "class":
          const escapedParts = expectation.parts.map((part) => {
            return Array.isArray(part)
              ? classEscape(part[0] as string) + "-" + classEscape(part[1] as string)
              : classEscape(part);
          });

          return "[" + (expectation.inverted ? "^" : "") + escapedParts + "]";
        case "any":
          return "any character";
        case "end":
          return "end of input";
        case "other":
          return expectation.description;
      }
    }

    function describeExpected(expected1: Expectation[]) {
      const descriptions = expected1.map(describeExpectation);
      let i: number;
      let j: number;

      descriptions.sort();

      if (descriptions.length > 0) {
        for (i = 1, j = 1; i < descriptions.length; i++) {
          if (descriptions[i - 1] !== descriptions[i]) {
            descriptions[j] = descriptions[i];
            j++;
          }
        }
        descriptions.length = j;
      }

      switch (descriptions.length) {
        case 1:
          return descriptions[0];

        case 2:
          return descriptions[0] + " or " + descriptions[1];

        default:
          return descriptions.slice(0, -1).join(", ")
            + ", or "
            + descriptions[descriptions.length - 1];
      }
    }

    function describeFound(found1: string | null) {
      return found1 ? "\"" + literalEscape(found1) + "\"" : "end of input";
    }

    return "Expected " + describeExpected(expected) + " but " + describeFound(found) + " found.";
  }

  public message: string;
  public expected: Expectation[];
  public found: string | null;
  public location: IFileRange;
  public name: string;

  constructor(message: string, expected: Expectation[], found: string | null, location: IFileRange) {
    super();
    this.message = message;
    this.expected = expected;
    this.found = found;
    this.location = location;
    this.name = "SyntaxError";

    if (typeof (Object as any).setPrototypeOf === "function") {
      (Object as any).setPrototypeOf(this, SyntaxError.prototype);
    } else {
      (this as any).__proto__ = SyntaxError.prototype;
    }
    if (typeof (Error as any).captureStackTrace === "function") {
      (Error as any).captureStackTrace(this, SyntaxError);
    }
  }

  format(sources: { source: string; text: string }[]): string {
    let str = 'Error: ' + this.message;
    if (this.location) {
      let src: string[] | null = null;
      let k;
      for (k = 0; k < sources.length; k++) {
        if (sources[k].source === this.location.source) {
          src = sources[k].text.split(/\r\n|\n|\r/g);
          break;
        }
      }
      let s = this.location.start;
      let loc = this.location.source + ':' + s.line + ':' + s.column;
      if (src) {
        let e = this.location.end;
        let filler = peg$padEnd('', s.line.toString().length, ' ');
        let line = src[s.line - 1];
        let last = s.line === e.line ? e.column : line.length + 1;
        str += '\n --> ' + loc + '\n' + filler + ' |\n' + s.line + ' | ' + line + '\n' + filler + ' | ' +
          peg$padEnd('', s.column - 1, ' ') +
          peg$padEnd('', last - s.column, '^');
      } else {
        str += '\n at ' + loc;
      }
    }
    return str;
  }
}

export interface ICached {
  nextPos: number;
  result: any;
}

function peg$parse(input: string, options?: IParseOptions) {
  options = options !== undefined ? options : {};

  const peg$FAILED: Readonly<any> = {};
  const peg$source = options.grammarSource;

  const peg$startRuleFunctions: {[id: string]: any} = { Formula: peg$parseFormula };
  let peg$startRuleFunction: () => any = peg$parseFormula;

  const peg$c0 = "|";
  const peg$c1 = peg$literalExpectation("|", false);
  const peg$c2 = "@";
  const peg$c3 = peg$literalExpectation("@", false);
  const peg$c4 = function(f1: any, fs: any): any {
        return fs.reduce((formula: Formula, arg: [string, string, string, Formula]) => {
          const [_1, op, _2, f] = arg;
          return { name: 'binary', op: opToName(op), children: [formula, f] };
        }, f1);
      };
  const peg$c5 = "&";
  const peg$c6 = peg$literalExpectation("&", false);
  const peg$c7 = "*";
  const peg$c8 = peg$literalExpectation("*", false);
  const peg$c9 = function(f1: any, fs: any): any {
      return fs.reduce((formula: Formula, arg: [string, string, string, Formula]) => {
          const [_1, op, _2, f] = arg;
          return { name: 'binary', op: opToName(op), children: [formula, f] };
        }, f1);
      };
  const peg$c10 = "~";
  const peg$c11 = peg$literalExpectation("~", false);
  const peg$c12 = "!";
  const peg$c13 = peg$literalExpectation("!", false);
  const peg$c14 = "?";
  const peg$c15 = peg$literalExpectation("?", false);
  const peg$c16 = function(op: any, f: any): any {
      return { name: 'unary', op: opToName(op), children: [f] };
    };
  const peg$c17 = function(f: any): any {
      return f;
    };
  const peg$c18 = /^[a-zA-Z]/;
  const peg$c19 = peg$classExpectation([["a", "z"], ["A", "Z"]], false, false);
  const peg$c20 = /^[a-zA-Z0-9]/;
  const peg$c21 = peg$classExpectation([["a", "z"], ["A", "Z"], ["0", "9"]], false, false);
  const peg$c22 = function(v: any): any {
      return {name: 'var', var: v[0] + v[1].join() };
    };
  const peg$c23 = "(";
  const peg$c24 = peg$literalExpectation("(", false);
  const peg$c25 = ")";
  const peg$c26 = peg$literalExpectation(")", false);
  const peg$c27 = peg$otherExpectation("whitespace");
  const peg$c28 = /^[ \t\n\r]/;
  const peg$c29 = peg$classExpectation([" ", "\t", "\n", "\r"], false, false);

  let peg$currPos = 0;
  let peg$savedPos = 0;
  const peg$posDetailsCache = [{ line: 1, column: 1 }];
  let peg$maxFailPos = 0;
  let peg$maxFailExpected: Expectation[] = [];
  let peg$silentFails = 0;

  const peg$resultsCache: {[id: number]: ICached} = {};

  let peg$result;

  if (options.startRule !== undefined) {
    if (!(options.startRule in peg$startRuleFunctions)) {
      throw new Error("Can't start parsing from rule \"" + options.startRule + "\".");
    }

    peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
  }

  function text(): string {
    return input.substring(peg$savedPos, peg$currPos);
  }

  function location(): IFileRange {
    return peg$computeLocation(peg$savedPos, peg$currPos);
  }

  function expected(description: string, location1?: IFileRange) {
    location1 = location1 !== undefined
      ? location1
      : peg$computeLocation(peg$savedPos, peg$currPos);

    throw peg$buildStructuredError(
      [peg$otherExpectation(description)],
      input.substring(peg$savedPos, peg$currPos),
      location1
    );
  }

  function error(message: string, location1?: IFileRange) {
    location1 = location1 !== undefined
      ? location1
      : peg$computeLocation(peg$savedPos, peg$currPos);

    throw peg$buildSimpleError(message, location1);
  }

  function peg$literalExpectation(text1: string, ignoreCase: boolean): ILiteralExpectation {
    return { type: "literal", text: text1, ignoreCase: ignoreCase };
  }

  function peg$classExpectation(parts: IClassParts, inverted: boolean, ignoreCase: boolean): IClassExpectation {
    return { type: "class", parts: parts, inverted: inverted, ignoreCase: ignoreCase };
  }

  function peg$anyExpectation(): IAnyExpectation {
    return { type: "any" };
  }

  function peg$endExpectation(): IEndExpectation {
    return { type: "end" };
  }

  function peg$otherExpectation(description: string): IOtherExpectation {
    return { type: "other", description: description };
  }

  function peg$computePosDetails(pos: number) {
    let details = peg$posDetailsCache[pos];
    let p;

    if (details) {
      return details;
    } else {
      p = pos - 1;
      while (!peg$posDetailsCache[p]) {
        p--;
      }

      details = peg$posDetailsCache[p];
      details = {
        line: details.line,
        column: details.column
      };

      while (p < pos) {
        if (input.charCodeAt(p) === 10) {
          details.line++;
          details.column = 1;
        } else {
          details.column++;
        }

        p++;
      }

      peg$posDetailsCache[pos] = details;

      return details;
    }
  }

  function peg$computeLocation(startPos: number, endPos: number): IFileRange {
    const startPosDetails = peg$computePosDetails(startPos);
    const endPosDetails = peg$computePosDetails(endPos);

    return {
      source: peg$source,
      start: {
        offset: startPos,
        line: startPosDetails.line,
        column: startPosDetails.column
      },
      end: {
        offset: endPos,
        line: endPosDetails.line,
        column: endPosDetails.column
      }
    };
  }

  function peg$fail(expected1: Expectation) {
    if (peg$currPos < peg$maxFailPos) { return; }

    if (peg$currPos > peg$maxFailPos) {
      peg$maxFailPos = peg$currPos;
      peg$maxFailExpected = [];
    }

    peg$maxFailExpected.push(expected1);
  }

  function peg$buildSimpleError(message: string, location1: IFileRange) {
    return new SyntaxError(message, [], "", location1);
  }

  function peg$buildStructuredError(expected1: Expectation[], found: string | null, location1: IFileRange) {
    return new SyntaxError(
      SyntaxError.buildMessage(expected1, found),
      expected1,
      found,
      location1
    );
  }

  function peg$parseFormula(): any {
    let s0, s1, s2, s3, s4, s5, s6, s7;

    const key = peg$currPos * 5 + 0;
    const cached: ICached = peg$resultsCache[key];

    if (cached) {
      peg$currPos = cached.nextPos;

      return cached.result;
    }

    s0 = peg$currPos;
    s1 = peg$parseFormula1();
    if (s1 as any !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      if (s4 as any !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 124) {
          s5 = peg$c0;
          peg$currPos++;
        } else {
          s5 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c1); }
        }
        if (s5 as any === peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 64) {
            s5 = peg$c2;
            peg$currPos++;
          } else {
            s5 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c3); }
          }
        }
        if (s5 as any !== peg$FAILED) {
          s6 = peg$parse_();
          if (s6 as any !== peg$FAILED) {
            s7 = peg$parseFormula1();
            if (s7 as any !== peg$FAILED) {
              s4 = [s4, s5, s6, s7];
              s3 = s4;
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 as any !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        if (s4 as any !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 124) {
            s5 = peg$c0;
            peg$currPos++;
          } else {
            s5 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c1); }
          }
          if (s5 as any === peg$FAILED) {
            if (input.charCodeAt(peg$currPos) === 64) {
              s5 = peg$c2;
              peg$currPos++;
            } else {
              s5 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c3); }
            }
          }
          if (s5 as any !== peg$FAILED) {
            s6 = peg$parse_();
            if (s6 as any !== peg$FAILED) {
              s7 = peg$parseFormula1();
              if (s7 as any !== peg$FAILED) {
                s4 = [s4, s5, s6, s7];
                s3 = s4;
              } else {
                peg$currPos = s3;
                s3 = peg$FAILED;
              }
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      if (s2 as any !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c4(s1, s2);
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }

    peg$resultsCache[key] = { nextPos: peg$currPos, result: s0 };

    return s0;
  }

  function peg$parseFormula1(): any {
    let s0, s1, s2, s3, s4, s5, s6, s7;

    const key = peg$currPos * 5 + 1;
    const cached: ICached = peg$resultsCache[key];

    if (cached) {
      peg$currPos = cached.nextPos;

      return cached.result;
    }

    s0 = peg$currPos;
    s1 = peg$parseFormula2();
    if (s1 as any !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      if (s4 as any !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 38) {
          s5 = peg$c5;
          peg$currPos++;
        } else {
          s5 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c6); }
        }
        if (s5 as any === peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 42) {
            s5 = peg$c7;
            peg$currPos++;
          } else {
            s5 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c8); }
          }
        }
        if (s5 as any !== peg$FAILED) {
          s6 = peg$parse_();
          if (s6 as any !== peg$FAILED) {
            s7 = peg$parseFormula2();
            if (s7 as any !== peg$FAILED) {
              s4 = [s4, s5, s6, s7];
              s3 = s4;
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 as any !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        if (s4 as any !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 38) {
            s5 = peg$c5;
            peg$currPos++;
          } else {
            s5 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c6); }
          }
          if (s5 as any === peg$FAILED) {
            if (input.charCodeAt(peg$currPos) === 42) {
              s5 = peg$c7;
              peg$currPos++;
            } else {
              s5 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c8); }
            }
          }
          if (s5 as any !== peg$FAILED) {
            s6 = peg$parse_();
            if (s6 as any !== peg$FAILED) {
              s7 = peg$parseFormula2();
              if (s7 as any !== peg$FAILED) {
                s4 = [s4, s5, s6, s7];
                s3 = s4;
              } else {
                peg$currPos = s3;
                s3 = peg$FAILED;
              }
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      if (s2 as any !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c9(s1, s2);
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }

    peg$resultsCache[key] = { nextPos: peg$currPos, result: s0 };

    return s0;
  }

  function peg$parseFormula2(): any {
    let s0, s1, s2, s3;

    const key = peg$currPos * 5 + 2;
    const cached: ICached = peg$resultsCache[key];

    if (cached) {
      peg$currPos = cached.nextPos;

      return cached.result;
    }

    s0 = peg$currPos;
    if (input.charCodeAt(peg$currPos) === 126) {
      s1 = peg$c10;
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c11); }
    }
    if (s1 as any === peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 33) {
        s1 = peg$c12;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c13); }
      }
      if (s1 as any === peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 63) {
          s1 = peg$c14;
          peg$currPos++;
        } else {
          s1 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c15); }
        }
      }
    }
    if (s1 as any !== peg$FAILED) {
      s2 = peg$parse_();
      if (s2 as any !== peg$FAILED) {
        s3 = peg$parseFormula2();
        if (s3 as any !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$c16(s1, s3);
          s0 = s1;
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 as any === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseFormula3();
      if (s1 as any !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c17(s1);
      }
      s0 = s1;
    }

    peg$resultsCache[key] = { nextPos: peg$currPos, result: s0 };

    return s0;
  }

  function peg$parseFormula3(): any {
    let s0, s1, s2, s3, s4, s5;

    const key = peg$currPos * 5 + 3;
    const cached: ICached = peg$resultsCache[key];

    if (cached) {
      peg$currPos = cached.nextPos;

      return cached.result;
    }

    s0 = peg$currPos;
    s1 = peg$currPos;
    if (peg$c18.test(input.charAt(peg$currPos))) {
      s2 = input.charAt(peg$currPos);
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c19); }
    }
    if (s2 as any !== peg$FAILED) {
      s3 = [];
      if (peg$c20.test(input.charAt(peg$currPos))) {
        s4 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s4 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c21); }
      }
      while (s4 as any !== peg$FAILED) {
        s3.push(s4);
        if (peg$c20.test(input.charAt(peg$currPos))) {
          s4 = input.charAt(peg$currPos);
          peg$currPos++;
        } else {
          s4 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c21); }
        }
      }
      if (s3 as any !== peg$FAILED) {
        s2 = [s2, s3];
        s1 = s2;
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
    } else {
      peg$currPos = s1;
      s1 = peg$FAILED;
    }
    if (s1 as any !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$c22(s1);
    }
    s0 = s1;
    if (s0 as any === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 40) {
        s1 = peg$c23;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c24); }
      }
      if (s1 as any !== peg$FAILED) {
        s2 = peg$parse_();
        if (s2 as any !== peg$FAILED) {
          s3 = peg$parseFormula();
          if (s3 as any !== peg$FAILED) {
            s4 = peg$parse_();
            if (s4 as any !== peg$FAILED) {
              if (input.charCodeAt(peg$currPos) === 41) {
                s5 = peg$c25;
                peg$currPos++;
              } else {
                s5 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c26); }
              }
              if (s5 as any !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$c17(s3);
                s0 = s1;
              } else {
                peg$currPos = s0;
                s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;
              s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }

    peg$resultsCache[key] = { nextPos: peg$currPos, result: s0 };

    return s0;
  }

  function peg$parse_(): any {
    let s0, s1;

    const key = peg$currPos * 5 + 4;
    const cached: ICached = peg$resultsCache[key];

    if (cached) {
      peg$currPos = cached.nextPos;

      return cached.result;
    }

    peg$silentFails++;
    s0 = [];
    if (peg$c28.test(input.charAt(peg$currPos))) {
      s1 = input.charAt(peg$currPos);
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c29); }
    }
    while (s1 as any !== peg$FAILED) {
      s0.push(s1);
      if (peg$c28.test(input.charAt(peg$currPos))) {
        s1 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c29); }
      }
    }
    peg$silentFails--;
    if (s0 as any === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c27); }
    }

    peg$resultsCache[key] = { nextPos: peg$currPos, result: s0 };

    return s0;
  }

  peg$result = peg$startRuleFunction();

  if (peg$result !== peg$FAILED && peg$currPos === input.length) {
    return peg$result;
  } else {
    if (peg$result !== peg$FAILED && peg$currPos < input.length) {
      peg$fail(peg$endExpectation());
    }

    throw peg$buildStructuredError(
      peg$maxFailExpected,
      peg$maxFailPos < input.length ? input.charAt(peg$maxFailPos) : null,
      peg$maxFailPos < input.length
        ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1)
        : peg$computeLocation(peg$maxFailPos, peg$maxFailPos)
    );
  }
}

export interface IParseOptions {
  filename?: string;
  startRule?: string;
  tracer?: any;
  [key: string]: any;
}
export type ParseFunction = (input: string, options?: IParseOptions) => any;
export const parse: ParseFunction = peg$parse;

