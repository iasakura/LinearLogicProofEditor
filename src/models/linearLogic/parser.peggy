{{
  import { Formula } from '../models/linearLogic/Formula';

  const opToName = (op: string) => {
    if (op == '|') { return 'or'; }
    if (op == '@') { return 'par'; }
    if (op == '*') { return 'tensor'; }
    if (op == '&') { return 'and'; }
    if (op == '~') { return 'not'; }
    if (op == '!') { return 'ofCourse'; }
    if (op == '?') { return 'whyNot'; }
  }
}}

Formula
  = f1:Formula1 fs:(_ op:("|" / "@") _ f2:Formula1)* {
      return fs.reduce((formula: Formula, arg: [string, string, string, Formula]) => {
        const [_1, op, _2, f] = arg;
        return { name: 'binary', op: opToName(op), children: [formula, f] };
      }, f1);
    }

Formula1
  = f1:Formula2 fs:(_ op:("&" / "*") _ f2:Formula2)* {
    return fs.reduce((formula: Formula, arg: [string, string, string, Formula]) => {
        const [_1, op, _2, f] = arg;
        return { name: 'binary', op: opToName(op), children: [formula, f] };
      }, f1);
    }

Formula2
  = op:("~" / "!" / "?") _ f:Formula2 {
    return { name: 'unary', op: opToName(op), children: [f] };
  } /
  f:Formula3 {
    return f;
  }

Formula3
  = v:([a-zA-Z] [a-zA-Z0-9]*) {
    return {name: 'var', var: v[0] + v[1].join() };
  } / "(" _ f:Formula _ ")" {
    return f;
  }

_ "whitespace"
  = [ \t\n\r]*
