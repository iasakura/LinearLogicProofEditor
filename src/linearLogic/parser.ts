import * as parser from '../../gen/parser';
import { Formula } from './Formula';

export const parse = (formula: string): Formula => {
  return parser.parse(formula) as Formula;
};
