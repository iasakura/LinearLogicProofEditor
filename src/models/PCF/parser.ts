import * as parser from '../../gen/pcf-parser';
import { Term } from './types';

export const parseTerm = (formula: string): Term => {
  return parser.parse(formula) as Term;
};
