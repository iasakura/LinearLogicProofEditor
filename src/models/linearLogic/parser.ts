import * as parser from '../../gen/parser';
import { Formula } from './Formula';

export const parseFormula = (formula: string): Formula => {
  return parser.parse(formula) as Formula;
};

export const parseSequent = (sequent: string): Formula[] => {
  return sequent
    .split(',')
    .filter((s) => s.length > 0)
    .map((s) => s.trim())
    .map((s) => parser.parse(s) as Formula);
};
