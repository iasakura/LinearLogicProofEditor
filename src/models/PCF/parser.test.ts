import { parseTerm } from './parser';

test('complex', () => {
  const input = 'lam f: nat -> nat-> nat. lam x: nat. f (f (f x))';
  const ret = parseTerm(input);
  expect(ret).toMatchSnapshot();
});
