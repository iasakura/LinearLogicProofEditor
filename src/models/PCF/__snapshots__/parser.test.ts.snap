// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex 1`] = `
Object {
  "body": Object {
    "body": Object {
      "name": "app",
      "terms": Array [
        Object {
          "name": "var",
          "v": "f",
        },
        Object {
          "name": "app",
          "terms": Array [
            Object {
              "name": "var",
              "v": "f",
            },
            Object {
              "name": "app",
              "terms": Array [
                Object {
                  "name": "var",
                  "v": "f",
                },
                Object {
                  "name": "var",
                  "v": "x",
                },
              ],
            },
          ],
        },
      ],
    },
    "name": "lambda",
    "typ": Object {
      "name": "nat",
    },
    "v": "x",
  },
  "name": "lambda",
  "typ": Object {
    "name": "arrow",
    "types": Array [
      Object {
        "name": "nat",
      },
      Object {
        "name": "arrow",
        "types": Array [
          Object {
            "name": "nat",
          },
          Object {
            "name": "nat",
          },
        ],
      },
    ],
  },
  "v": "f",
}
`;
